stlab/concurrency/utility.hpp

```diff
@@ -99,32 +118,110 @@ T blocking_get(future<T> x) {
     return std::move(*result);
 }
 
+template <typename T>
+stlab::optional<T> blocking_get(future<T> x, const std::chrono::nanoseconds& timeout) {
+    std::exception_ptr error = nullptr;
+    auto state = std::make_shared<detail::shared_state_result<T>>();
+    auto hold = std::move(x).recover(immediate_executor, [_weak_state = make_weak_ptr(state)](auto&& r) {
+        auto state = _weak_state.lock();
+        if (!state) {
+            return;
+        }
+        if (r.error())
+            state->error = *std::forward<decltype(r)>(r).error();
+        else
+            state->result = std::move(*std::forward<decltype(r)>(r).get_try());
+        
+        {
+            std::unique_lock<std::mutex> lock{state->m};
+            state->flag = true;
+            /*
+                WARNING : Calling `notify_one()` inside the lock is a pessimization because
+                it means the code waiting will block aquiring the lock as soon as it wakes up.
+
+                However, if we do the notificiation outside the lock, blocking_get will return
+                and we'll be calling a destructed condition variable.
+            */
```

Comment:
As soon as we have s shared_state, the statement in the comment is not true any more. Since the condition_variable IS in the shared_state, it is guaranteed that it is not destructed. So in this case, one could remove the pessimization by moving the notify_one outside of the scope (and remove the comment :-) )
This is true for both cases.